### ISSUE TITLE
DNS Hijacking UI Protection

### ISSUE THREAD

**Issue by 0x4007** [pavlovcik/uad-ui#343]
Perhaps a more practical solution would be to compare the UI with this repo's latest `production` commit. 

There might be a means to hash the entire UI codebase and match the hash with some request on the frontend. This probably requires just a little bit more research before creating this bounty, but I think this is a relatively simple to implement layer of protection because this repository is open sourced.

> Since DNS hijacking is on the rise (we've seen it recently with the curve.fi UI being compromised) an idea would be to have a badge on the UI that does an ajax request to get the commit hash of the current instance (`uad.ubq.fi`) and compare with the other deployments (`netlify.ubq.fi`, `fleek.ubq.fi`, `cloudflare.ubq.fi`) 

> This would require a compromise of all UI deployments simultaneously to thwart this detection.

> If there's any difference then we can get the user's attention and explain that something is fishy and to be aware of any new approvals!

_Originally posted by @pavlovcik in https://github.com/ubiquity/ubiquity-dollar/discussions/197#discussioncomment-3647038_

**Comment by rndquu** [comment]
So the idea is the following:
1. `uad.ubq.fi` is loaded
2. js script from `uad.ubq.fi` calculates the UI hash
3. js script from `uad.ubq.fi` gets UI hashes from `netlify.ubq.fi`, `fleek.ubq.fi` and `cloudflare.ubq.fi `
4. if UI hash from `uad.ubq.fi` not equal to at least one UI hash from `netlify.ubq.fi, fleek.ubq.fi, cloudflare.ubq.fi` then show a warning to the user

Questions:
1. Is the example correct?
2. The attacker has full access to the frontend in case of DNS hijacking so he would be able to hide this warning or mitigate this "UI hash check" in any other way. Is it ok?

**Comment by 0x4007** [comment]
> The attacker has full access to the frontend in case of DNS hijacking so he would be able to hide this warning or mitigate this "UI hash check" in any other way. Is it ok?

Any ideas around this? I know the solution isn't perfect but I can't think of anything much better at this time. Perhaps we can just disable `approval` requests in an obfuscated manner to make it more tedious for hackers.

**Comment by rndquu** [comment]
we could do both: warning message + web3 blocking (like approval requests and anything else)

anyway you can assign it to me, I will let you know if any other idea comes to mind

**Comment by 0x4007** [comment]
What if we use [wasm](https://webassembly.org/)? I heard that its sort of like a blackbox. Perhaps there is a way to include it to act as sort of a "key" to enable the UI.

**Comment by rndquu** [comment]
> What if we use [wasm](https://webassembly.org/)? I heard that its sort of like a blackbox. Perhaps there is a way to include it to act as sort of a "key" to enable the UI.

I'll check it

**Comment by 0x4007** [comment]
New idea: there was a famous example floating around the internet some years back with an XSS example with a picture of a squirrel. I can't find it now, but they were able to embed code inside the squirrel image to render HTML. But when you view the source code of the HTML page it isn't there, and there isn't any JavaScript as far as I remember. What if we consider XSS as a strategy to hide code?

---

Your thoughtful writeup on the PR along with the solution implemented is sufficient to collect the bounty though. What is your DAI address?

**Comment by rndquu** [comment]
> New idea: there was a famous example floating around the internet some years back with an XSS example with a picture of a squirrel. I can't find it now, but they were able to embed code inside the squirrel image to render HTML. But when you view the source code of the HTML page it isn't there, and there isn't any JavaScript as far as I remember. What if we consider XSS as a strategy to hide code?
> 
> Your thoughtful writeup on the PR along with the solution implemented is sufficient to collect the bounty though. What is your DAI address?

Yes, I tried the "XSS way", trying to include js in svg, img, meta tags but next js is doing a really good job removing all malicious scripts for production builds (even when they work in development environment they still get removed on production build).

There's also the way to modify webpack config(for example to include img tag with XSS insude with "DNS hijack prevention" script), but next js says in their docs that you shouldn't modify webpack config as next js can simply update something in the future and your script may silently stop working.

Then consider the "attacker way". Once he sees a huge red warning it will take him literally a minute to find out what is causing it no matter how you're hiding the script.

> But when you view the source code of the HTML page it isn't there

Yes, you can add js script dynamically. Like put your js logic in an external script, and then on page load dynamically add new script tag with source to that external script. This way it is really hard to find the script via chrome's "view source" feature as this is a pretty hard task for chrome. But when the attacker clones the website's frontend and runs website on localhost he can easily see the dynamically injected code in the sources.

Simply put, you can not implement such a protection on the UI side, only make the attacker to spend few more minutes to find out what's causing a big red warning.

**Comment by rndquu** [comment]
address: 0x336C033842FA316d470e820c81b742e62A0765DC

**Comment by 0x4007** [comment]
> address: 0x336C033842FA316d470e820c81b742e62A0765DC

https://etherscan.io/tx/0x0ca4ff573b080c37d7e323949e6f1529738855c4ffc962156dd33b2b5f3cf868
https://etherscan.io/tx/0x1d57968145060e38c14715ecbf96494643c51097c83521d472e654ecd5cf8f9d
